src/main/java/org/magicdgs/readtools/utils/mappability/gem/GemMappabilityReader.java

```diff
@@ -0,0 +1,304 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.magicdgs.readtools.utils.mappability.gem;
+
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.util.BufferedLineReader;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.RuntimeIOException;
+import org.apache.commons.lang3.Range;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Iterator over gem-mappability number of mappings.
+ *
+ * <p>The GEM-mappability FASTA-like format consists on:
+ * <ul>
+ *     <li>Header with metadata. Includes information about the parameters used and the encoding for
+ *     the values of mappability. See {@link GemMappabilityHeader} for more information</li>
+ *     <li>Per-base range of values. Encoded as a char, representing a range of number of mappings.</li>
+ * </ul>
+ *
+ * <p>This class iterates over the file and retrieves the per-base range of values as a
+ * {@link GemMappabilityRecord}, which contains the sequence name, position and range of values.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public final class GemMappabilityReader implements CloseableIterator<GemMappabilityRecord> {
+
+    // HEADER PREFIX (for sequence name and meta-data)
+    private static final char HEADER_PREFIX = '~';
+
+    ///////////////////////////
+    // HEADER LINES - prefixed by two HEADER_PREFIX
+
+    private static final String KMER_LENGTH_HEADER_TEXT = "K-MER LENGTH";
+    private static final String APPROX_THRESHOLD_HEADER_TEXT = "APPROXIMATION THRESHOLD";
+    private static final String MAX_MISMATCH_HEADER_TEXT = "MAX MISMATCHES";
+    private static final String MAX_ERRORS_HEADER_TEXT = "MAX ERRORS";
+    private static final String MAX_INDEL_LENGTH_HEADER_TEXT = "MAX BIG INDEL LENGTH";
+    private static final String MIN_MATCH_HEADER_TEXT = "MIN MATCHED BASES";
+    private static final String STRATA_AFTER_BEST_HEADER_TEXT = "STRATA AFTER BEST";
+    private static final String ENCODING_HEADER_TEXT = "ENCODING";
+
+    // pattern for the encoding (char to value range)
+    protected static final Pattern ENCODING_PATTERN = Pattern.compile("'(.)'~\\[(\\d+)-(\\d+)]");
```

Comment:
Why protected? Add a comment and if it is just for testing include the `@VisibleForTesting`